state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,214 @@
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use serde::{Deserialize, Serialize};
+use sp_std::collections::{btree_map::BTreeMap, btree_set::BTreeSet};
+
+#[cfg(feature = "runtime-benchmarks")]
+use cf_chains::benchmarking_value::BenchmarkValue;
+
+use crate::{
+	electoral_system::{
+		AuthorityVoteOf, ConsensusVotes, ElectionReadAccess, ElectionWriteAccess, ElectoralSystem,
+		ElectoralWriteAccess, VotePropertiesOf,
+	},
+	vote_storage::{self, VoteStorage},
+	CorruptStorageError, ElectionIdentifier,
+};
+use cf_chains::sol::{
+	api::ContractSwapAccountAndSender, MAX_BATCH_SIZE_OF_CONTRACT_SWAP_ACCOUNT_CLOSURES,
+	MAX_WAIT_BLOCKS_FOR_SWAP_ACCOUNT_CLOSURE_APICALLS,
+	NONCE_AVAILABILITY_THRESHOLD_FOR_INITIATING_SWAP_ACCOUNT_CLOSURES,
+};
+use cf_utilities::success_threshold_from_share_count;
+use frame_support::{
+	pallet_prelude::{MaybeSerializeDeserialize, Member},
+	sp_runtime::traits::CheckedSub,
+	Parameter,
+};
+use itertools::Itertools;
+use sp_std::vec::Vec;
+
+pub trait SolanaVaultSwapAccountsHook<Account, SwapDetails, E> {
+	fn close_accounts(accounts: Vec<Account>) -> Result<(), E>;
+	fn initiate_vault_swap(swap_details: SwapDetails);
+	fn get_number_of_available_sol_nonce_accounts() -> usize;
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize, TypeInfo, Encode, Decode)]
+pub struct SolanaVaultSwapsElectoralState<Account: Ord, BlockNumber> {
+	pub block_number_last_closed_accounts: BlockNumber,
+	pub witnessed_open_accounts: Vec<Account>,
+	pub closure_initiated_accounts: BTreeSet<Account>,
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+impl BenchmarkValue for SolanaVaultSwapsElectoralState<ContractSwapAccountAndSender, u32> {
+	fn benchmark_value() -> Self {
+		Self {
+			block_number_last_closed_accounts: 1u32,
+			witnessed_open_accounts: vec![BenchmarkValue::benchmark_value()],
+			closure_initiated_accounts: BTreeSet::from([BenchmarkValue::benchmark_value()]),
+		}
+	}
+}
+
+#[derive(
+	Clone, PartialEq, Eq, Debug, Serialize, Deserialize, TypeInfo, Encode, Decode, Ord, PartialOrd,
+)]
+pub struct SolanaVaultSwapsVote<Account: Ord, SwapDetails: Ord> {
+	pub new_accounts: BTreeSet<(Account, SwapDetails)>,
+	pub confirm_closed_accounts: BTreeSet<Account>,
+}
+
+pub struct SolanaVaultSwapAccounts<
```

Comment:
there are no tests?

Reply:
wanted to first nail down the approach based on the reviews before writing any tests

Reply:
working on unit tests right now. @albert-llimos added some end to end tests in the bouncer.